### ISSUE TITLE
Explore removing unused zookeeper watchers

### ISSUE THREAD

**Issue by keith-turner** [apache/accumulo#1220]
For per table config, Accumulo creates zookeeper watches per table per tablet servers.  When a table is deleted these watches may be left behind consuming memory on the ZK servers.  Accumulo may be able to use the following to remove these watchers.

https://issues.apache.org/jira/browse/ZOOKEEPER-1829

**Comment by ctubbsii** [comment]
If this can be done with low-risk and a small amount of code, it could be backported to 2.0.x or even 1.9.x; but without trying, it's hard to know for sure.

I think the basic design would be something like:

1. Have ZooCache keep references to any Watchers,
2. ZooCache implements a method to match any watchers against some predicate and delete them,
3. In the Master, when a table is deleted, it triggers deletion of any watchers matching paths to that table's location in ZK, and
4. In the TServers, trigger deletion of any watchers for tables that don't match any loaded tablets, whenever a tablet is unloaded or periodically.

**Comment by jzgithub1** [comment]
I am working on this.

**Comment by jzgithub1** [comment]
@keith-turner  and @ctubbsii ,  from the basic design above I will try to implement number 2:
ZooCache implements a method to match any watchers against some predicate and delete them.
That is the least risky and fastest path in my opinion.   I will implement the other steps afterword.

**Comment by ctubbsii** [comment]
@jzgithub1 The above design will only work if we upgrade to the latest ZK 3.5. There may be other implications of that, so that should be done first, as its own task, before implementing the above.

An alternative might be to simply use fewer watchers in the first place, by changing the design for table config updates to use a single global watcher on a single field, named something like "/configChanged".

**Comment by jzgithub1** [comment]
Thanks @ctubbsii .  Glad I didn't go down that road too far.

**Comment by ctubbsii** [comment]
To be clear, I don't know for sure, which methods are our best bets in terms of simplifying the code or reducing the number of watchers. It may require some investigation, and prototyping of the various options. I just wanted to make sure whoever worked on this was aware that there are other possible solutions.

**Comment by ctubbsii** [comment]
This is superseded by #1454
